feat:add-objectTrackerOption-fix-channelFull-panic

ryansxy opened the pull request:
#### What type of PR is this?

/kind feature


#### What this PR does / why we need it:

Because of using fake client The program panics with "channel full", panic may not be the best solution, sometimes we don't want the program to panic when it have fake client.



#### Which issue(s) this PR fixes:

Fixes #116700 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

ryansxy committed:
feat:add-objectTrackerOption-fix-channelFull-panic

Signed-off-by: ryansxy <2903326670@qq.com>

cici37 commented:
/assign @dinhxuanvu
Thank you!
/triage accepted

dinhxuanvu commented on staging/src/k8s.io/client-go/kubernetes/fake/clientset_generated.go:
```diff
@@ -155,6 +155,33 @@ func NewSimpleClientset(objects ...runtime.Object) *Clientset {
 
+// NewSimpleClientsetWithObjectTrackerOption returns a clientset that will respond with the provided objects.
+// its function is consistent with NewSimpleClientset, the difference is that NewSimpleClientsetWithObjectTrackerOption
+// will customize clientSet.ObecjectTracker, the current optional is custom watcher
```
Just a minor type here. `clientSet.ObecjectTracker` to `clientSet.ObjectTracker`.

ryansxy replied:
Ok, thank you, I have edited it

dinhxuanvu reviewed (commented):
Looking good to me. I don't have permission to enable the testing.

dinhxuanvu commented:
@cici37 Would you mind adding the label ok-to-test this PR? Thanks

cici37 commented:
Sure!
/ok-to-test

ryansxy commented:
@cici37 @dinhxuanvu 
When running the hack/update-codegen.sh script, I found that I changed the staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_clientset.go file, but it was actually generated by code-generator, so I made some adjustments, and I added the NewSimpleClientsetWithObjectTrackerOption method in code-generator，thank you for your help

ryansxy commented:
/check-cla

ryansxy commented:
/easycla

dinhxuanvu reviewed (commented):
/lgtm

deads2k commented:
Everybody love panics in the test code :)

/lgtm
/approve

aojea commented:
also, can we have a test that makes it panic to demonstrate this solves the panic?


/hold

gjkim42 commented:
/cc
